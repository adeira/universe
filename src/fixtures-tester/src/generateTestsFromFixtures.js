// @flow

import os from 'os';
import { isObject } from '@adeira/js';

const FIXTURE_TAG = Symbol.for('FIXTURE_TAG');

/**
 * Extend Jest with a custom snapshot serializer to provide additional context
 * and reduce the amount of escaping that occurs. This serializer is heavily
 * inspired by Relay test cases.
 */
expect.addSnapshotSerializer({
  print(serializerValue) {
    return Object.keys(serializerValue)
      .map((key) => {
        const value = serializerValue[key];
        const inspectedValue = isObject(value) ? JSON.stringify(value, null, 2) : value;
        return `~~~~~~~~~~ ${key.toUpperCase()} ~~~~~~~~~~\n${inspectedValue}`;
      })
      .join(os.EOL);
  },
  test(value) {
    return value && value[FIXTURE_TAG] === true;
  },
});

type OperationOutput = any | Promise<any>;

/**
 * Generates a set of jest snapshot tests that compare the output of the
 * provided `operation` to each of the matching files in the `fixturesPath`.
 */
export default function generateTestsFromFixtures( // eslint-disable-line jest/no-export
  fixturesPath: string,
  operation: (input: string) => OperationOutput,
  snapshotName?: string,
): void {
  const fs = require('fs');
  const path = require('path');

  let fixtures = fs.readdirSync(fixturesPath);

  test(`has fixtures in ${fixturesPath}`, () => {
    expect(fixtures.length > 0).toBe(true);
  });

  const shouldSkip = (file) => /\.only\.\w+$/.test(file);
  const isFile = (f) => fs.lstatSync(path.join(fixturesPath, f)).isFile();
  const onlyFixtures = fixtures.filter(shouldSkip);
  if (onlyFixtures.length) {
    /* eslint-disable jest/no-disabled-tests */
    // $FlowFixMe[prop-missing]: we need to update our Jest type definitions (TODO)
    test.skip.each(fixtures.filter((name) => !shouldSkip(name)))(
      'matches expected output: %s',
      () => {},
    );
    /* eslint-enable jest/no-disabled-tests */
    fixtures = onlyFixtures;
  }

  // eslint-disable-next-line jest/no-identical-title
  test.each(fixtures.filter(isFile))('matches expected output: %s', async (file) => {
    const input = fs.readFileSync(path.join(fixturesPath, file), 'utf8');
    const output = await getOutputForFixture(input, operation, file);
    if (snapshotName != null) {
      expect({
        // $FlowIssue[invalid-computed-prop]: https://github.com/facebook/flow/issues/3258
        [FIXTURE_TAG]: true,
        input: input,
        output: output,
      }).toMatchSnapshot(snapshotName);
    } else {
      expect({
        // $FlowIssue[invalid-computed-prop]: https://github.com/facebook/flow/issues/3258
        [FIXTURE_TAG]: true,
        input: input,
        output: output,
      }).toMatchSnapshot();
    }
  });
}

async function getOutputForFixture(
  input: string,
  operation: (input: string) => OperationOutput,
  file: string,
): Promise<any> {
  const shouldThrow = /\.error\.\w+$/.test(file);
  if (shouldThrow) {
    let result;
    try {
      result = await operation(input);
    } catch (error) {
      return `THROWN EXCEPTION:\n\n${error.toString()}`;
    }
    throw new Error(`Expected test file '${file}' to throw, but it passed:\n${result}`);
  } else {
    return operation(input);
  }
}
