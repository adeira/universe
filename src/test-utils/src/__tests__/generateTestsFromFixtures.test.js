// @flow

import path from 'path';

import generateTestsFromFixtures from '../generateTestsFromFixtures';

generateTestsFromFixtures(path.join(__dirname, `/fixtures/simple`), (input) => input.toUpperCase());
generateTestsFromFixtures(path.join(__dirname, `/fixtures/simple`), (input) =>
  Promise.resolve(input.toUpperCase()),
);

generateTestsFromFixtures(
  path.join(__dirname, `/fixtures/simple`),
  (input) => input,
  'custom snapshot name (no modifications)',
);

generateTestsFromFixtures(path.join(__dirname, `/fixtures/simple`), (input) => {
  return {
    payload: JSON.stringify(input),
    // testing pretty object printing
    p: { r: { e: { t: { t: { y: true } } } } },
  };
});

generateTestsFromFixtures(path.join(__dirname, `/fixtures/errors`), () => {
  throw new Error('correctly failed');
});
