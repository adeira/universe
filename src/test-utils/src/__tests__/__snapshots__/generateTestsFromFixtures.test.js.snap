// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches expected output: aaa.error.txt 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
aaa error

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
THROWN EXCEPTION:

Error: correctly failed
`;

exports[`matches expected output: aaa.txt 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
aaa

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
AAA

`;

exports[`matches expected output: aaa.txt 2`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
aaa

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
AAA

`;

exports[`matches expected output: aaa.txt 3`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
aaa

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
{
  "payload": "\\"aaa\\\\n\\"",
  "p": {
    "r": {
      "e": {
        "t": {
          "t": {
            "y": true
          }
        }
      }
    }
  }
}
`;

exports[`matches expected output: aaa.txt: custom snapshot name (no modifications) 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
aaa

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
aaa

`;

exports[`matches expected output: bbb.txt 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
bbb

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
BBB

`;

exports[`matches expected output: bbb.txt 2`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
bbb

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
BBB

`;

exports[`matches expected output: bbb.txt 3`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
bbb

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
{
  "payload": "\\"bbb\\\\n\\"",
  "p": {
    "r": {
      "e": {
        "t": {
          "t": {
            "y": true
          }
        }
      }
    }
  }
}
`;

exports[`matches expected output: bbb.txt: custom snapshot name (no modifications) 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
bbb

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
bbb

`;
